File: core/app/src/main/java/fr/sncf/osrd/envelope_sim_infra/MRSP.java
package fr.sncf.osrd.envelope_sim_infra;

import static java.lang.Math.max;
import static java.lang.Math.min;

import fr.sncf.osrd.envelope.*;
import fr.sncf.osrd.envelope.part.EnvelopePart;
import fr.sncf.osrd.envelope_sim.EnvelopeProfile;
import fr.sncf.osrd.infra.trackgraph.TrackSection;
import fr.sncf.osrd.train.RollingStock;
import fr.sncf.osrd.train.TrainPath;
import fr.sncf.osrd.utils.DoubleUtils;
import fr.sncf.osrd.utils.graph.EdgeDirection;
import java.util.List;

/** MRSP = most restrictive speed profile: maximum speed allowed at any given point */
public class MRSP {
    public enum LimitKind implements EnvelopeAttr {
        SPEED_LIMIT,
        TRAIN_LIMIT,
        ;

        @Override
        public Class<? extends EnvelopeAttr> getAttrType() {
            return LimitKind.class;
        }
    }

    /** Computes the most restricted speed profile */
    public static Envelope from(TrainPath trainPath, RollingStock rollingStock) {
        var builder = new MRSPEnvelopeBuilder();

        // add a limit for the maximum speed the hardware is rated for
        builder.addPart(EnvelopePart.generateTimes(
                List.of(EnvelopeProfile.CONSTANT_SPEED, LimitKind.TRAIN_LIMIT),
                new double[] { 0, trainPath.length },
                new double[] { rollingStock.maxSpeed, rollingStock.maxSpeed }
        ));

        var offset = 0;
        for (var trackSectionRange : trainPath.trackSectionPath) {
            var edge = trackSectionRange.edge;
            for (var speedRange : TrackSection.getSpeedSections(edge, trackSectionRange.direction)) {
                // ignore the speed limit if it doesn't apply to our train
                var speedSection = speedRange.value;
                if (!speedSection.isValidFor(rollingStock))
                    continue;

                var speedRangeBegin = min(speedRange.begin, speedRange.end);
                var speedRangeEnd = max(speedRange.begin, speedRange.end);
                var beginOnRange = 0.;
                if (trackSectionRange.direction == EdgeDirection.START_TO_STOP) {
                    beginOnRange = DoubleUtils.clamp(
                            speedRangeBegin - trackSectionRange.getBeginPosition(), 0, trackSectionRange.length());
                } else {
                    beginOnRange = DoubleUtils.clamp(
                            trackSectionRange.getBeginPosition() - speedRangeEnd, 0, trackSectionRange.length());
                }
                var endOnRange = DoubleUtils.clamp(
                        beginOnRange + speedRangeEnd - speedRangeBegin, 0, trackSectionRange.length());

                if (Double.compare(beginOnRange, endOnRange) == 0)
                    continue;

                // compute where this limit is active from and to
                var begin = offset + beginOnRange;
                var end = offset + endOnRange;

                // Add the envelope part corresponding to the restricted speed section
                builder.addPart(EnvelopePart.generateTimes(
                        List.of(EnvelopeProfile.CONSTANT_SPEED, LimitKind.SPEED_LIMIT),
                        new double[] { begin, end },
                        new double[] { speedSection.speedLimit, speedSection.speedLimit }
                ));
            }
            offset += trackSectionRange.length();
        }
        return builder.build();
    }
}

File: core/src/main/kotlin/fr/sncf/osrd/envelope_sim_infra/MRSP.kt
package fr.sncf.osrd.envelope_sim_infra

import fr.sncf.osrd.envelope.Envelope
import fr.sncf.osrd.envelope.MRSPEnvelopeBuilder
import fr.sncf.osrd.envelope.part.EnvelopePart
import fr.sncf.osrd.envelope_sim.EnvelopeProfile
import fr.sncf.osrd.envelope_sim.PhysicsRollingStock
import fr.sncf.osrd.sim_infra.api.PathProperties
import fr.sncf.osrd.utils.SelfTypeHolder
import fr.sncf.osrd.utils.rangeMapEntryToSpeedLimitProperty
import fr.sncf.osrd.utils.units.Distance.Companion.toMeters
import fr.sncf.osrd.utils.units.Speed.Companion.toMetersPerSecond
import kotlin.math.min

/**
 * Computes the MSRP for a rolling stock on a given path. (MRSP = most restrictive speed profile:
 * maximum speed allowed at any given point)
 *
 * @param path corresponding path.
 * @param rollingStock corresponding rolling stock.
 * @param addRollingStockLength whether the rolling stock length should be taken into account in the
 *   computation.
 * @param trainTag corresponding train.
 * @return the corresponding MRSP as an Envelope.
 */
fun computeMRSP(
    path: PathProperties,
    rollingStock: PhysicsRollingStock,
    addRollingStockLength: Boolean,
    trainTag: String?
): Envelope {
    return computeMRSP(
        path,
        rollingStock.getMaxSpeed(),
        rollingStock.getLength(),
        addRollingStockLength,
        trainTag
    )
}

/**
 * Computes the MSRP for a rolling stock on a given path.
 *
 * @param path corresponding path.
 * @param rsMaxSpeed rolling stock max speed (m/s)
 * @param rsLength length of the rolling stock (m)
 * @param addRollingStockLength whether the rolling stock length should be taken into account in the
 *   computation.
 * @param trainTag corresponding train.
 * @return the corresponding MRSP as an Envelope.
 */
fun computeMRSP(
    path: PathProperties,
    rsMaxSpeed: Double,
    rsLength: Double,
    addRollingStockLength: Boolean,
    trainTag: String?
): Envelope {
    val builder = MRSPEnvelopeBuilder()
    val pathLength = toMeters(path.getLength())

    // Add a limit corresponding to the hardware's maximum operational speed
    builder.addPart(
        EnvelopePart.generateTimes(
            listOf(EnvelopeProfile.CONSTANT_SPEED, MRSPEnvelopeBuilder.LimitKind.TRAIN_LIMIT),
            doubleArrayOf(0.0, pathLength),
            doubleArrayOf(rsMaxSpeed, rsMaxSpeed)
        )
    )

    val offset = if (addRollingStockLength) rsLength else 0.0
    val speedLimitProperties = path.getSpeedLimitProperties(trainTag)
    for (speedLimitPropertyRange in speedLimitProperties) {
        // Compute where this limit is active from and to
        val start = toMeters(speedLimitPropertyRange.lower)
        val end = min(pathLength, offset + toMeters(speedLimitPropertyRange.upper))
        val speedLimitProp = rangeMapEntryToSpeedLimitProperty(speedLimitPropertyRange)
        val speed = toMetersPerSecond(speedLimitProp.speed)
        val attrs =
            mutableListOf<SelfTypeHolder?>(
                EnvelopeProfile.CONSTANT_SPEED,
                MRSPEnvelopeBuilder.LimitKind.SPEED_LIMIT
            )
        if (speedLimitProp.source != null) {
            attrs.add(speedLimitProp.source)
        }
        if (speed != 0.0) // Add the envelope part corresponding to the restricted speed section
            builder.addPart(
                EnvelopePart.generateTimes(
                    attrs,
                    doubleArrayOf(start, end),
                    doubleArrayOf(speed, speed)
                )
            )
    }
    return builder.build()
}
